src/main.cpp

```diff
@@ -1224,6 +1335,162 @@ class ClaimValidator
             FormatMoney(m_claim.m_research_subsidy),
             FormatMoney(m_claim.m_block_subsidy)));
     }
+
+    // Cf. CreateMRCRewards which is this method's conjugate. Note the parameters are out parameters.
+    //
+    // Note that it is possible that someone could try and circumvent the 100% fee penalty for submitting an MRC within
+    // the zero payout interval by purposefully miscomputing the fee. This is one of the reasons why the fee is
+    // recomputed by the checking node as part of the ValidateMRC function. There are a number of checks done here to
+    // detect possible abuses of the coinstake and ensure that the MRC portion of the coinstake corresponds to the claim.
+    bool CheckMRCRewards(CAmount& mrc_rewards, CAmount& mrc_staker_fees, CAmount& mrc_fees) const
+    {
+        // For convenience
+        const CTransaction& coinstake = m_block.vtx[1];
+
+        unsigned int mrc_outputs = 0;
+        unsigned int mrc_output_limit = GetMRCOutputLimit(m_block.nVersion, false);
+        unsigned int mrc_claimed_outputs = m_claim.m_mrc_tx_map.size();
+
+        // If the number of MRCs in the claim's mrc tx map exceeds the output limit, then validation fails. This would be
+        // a problem introduced by the staking node and should get the same DoS as other claim validation errors above.
+        if (mrc_claimed_outputs > mrc_output_limit) {
+            return m_block.DoS(10, error(
+                                   "ConnectBlock[%s]: MRC claimed outputs, %u, exceeds max of %u excluding foundation "
+                                   "sidestake.",
+                                   __func__,
+                                   mrc_claimed_outputs,
+                                   mrc_output_limit));
+        }
+
+        if (mrc_output_limit > 0) {
+            Fraction foundation_fee_fraction = FoundationSideStakeAllocation();
+
+            for (const auto& tx : m_block.vtx) {
+                for (const auto& mrc : m_claim.m_mrc_tx_map) {
+                    if (mrc.second == tx.GetHash()) {
+                        for (const auto& contract : tx.GetContracts()) {
+                            // We only are processing MRC contracts here in this loop.
+                            if (contract.m_type != GRC::ContractType::MRC) continue;
+
+                            GRC::MRC mrc = contract.CopyPayloadAs<GRC::MRC>();
+
+                            if (const GRC::CpidOption cpid = mrc.m_mining_id.TryCpid()) {
+                                CBlockIndex* mrc_index = mapBlockIndex[mrc.m_last_block_hash];
+
+                                const GRC::BeaconOption beacon = GRC::GetBeaconRegistry().TryActive(*cpid, mrc_index->nTime);
+
+                                if (beacon) {
```

Comment:
TrySignMRC will prevent expired beacons to send, should people with expired beacons be able to make MRC requests for the last time to claim their rewards?

Reply:
I don't think so. The accrual will bridge an expired beacon if someone advertises and validates a new beacon using gthe same CPID. The key will be different and they will have to send a new MRC request, but I think the strict approach is safer.